metals/src/main/scala/scala/meta/internal/parsing/TokenEditDistance.scala

```diff
@@ -355,10 +363,16 @@ object TokenEditDistance {
   /**
    * Compare tokens only by their text and token category.
    */
-  private object TokenEqualizer extends Equalizer[Token] {
-    override def equals(original: Token, revised: Token): Boolean =
-      original.productPrefix == revised.productPrefix &&
-        original.pos.text == revised.pos.text
+  private object TokenEqualizer extends Equalizer[BaseToken] {
+    override def equals(original: BaseToken, revised: BaseToken): Boolean =
+      (original, revised) match {
+        case (ScalaToken(org), ScalaToken(rev)) =>
+          org.productPrefix == rev.productPrefix &&
+            org.pos.text == rev.pos.text
+        case (org: JavaToken, rev: JavaToken) =>
+          org.id == rev.id && org.text == rev.text
+        case _ => false
+      }
```

Comment:
I tried, but the Equalizer is invariant, which is causing  it not to compile :/

Reply:
As I see it used only in `fromTokens`. 
I mean smth like that:
```scala
private def fromTokens[A <: BaseToken](
  original: Seq[A],
  revised: Seq[A],
  equalizer: TokenEqualizer[A]
): TokenEditDistance =
  ...
  
def apply(
      originalInput: Input.VirtualFile,
      revisedInput: Input.VirtualFile,
      trees: Trees,
      doNothingWhenUnchanged: Boolean = true
  ): TokenEditDistance = {
     .. 
     
     if (isJava) fromTokens(..., ..., JavaTokenEqualizer)
     else  fromTokens(..., ..., ScalaTokenEqualizer)
 }

private object ScalaTokenEqualizer extends Equalizer[Token] {
  ...
} 
private object JavaTokenEqualizer extends Equalizer[JavaToken] {
  ...
}
 
```

Reply:
That works! But we still need to have the additional ScalaToken though :thinking:  So we need to wrap all the scalameta tokens. We could probably use structural types, but that feels a bit much :sweat_smile:

Reply:
Or maybe it's not that bad? Take a look at the current state (unless I broke some more tests now :scream: )

Reply:
I never checked that by myself but as I know structural types are using reflection that adds overhead.

I think the best would be to introduce a type class for working with any type of token. But it would require more refactoring in `TokenEditDistance` but that way should keep the performance as it is.
Smth like:
```scala
trait TokenOps[A] {
  def pos(a: A): Position
  def input(a: A): Input
  def start(a: A): Int
  def end(a: A): Int
}

object TokenOps {

  implicit val scalaToken: TokenOps[Token] =
    new TokenOps[Token] {
      def pos(a: Token): Position = a.pos
      def input(a: Token): Input = a.input
      def start(a: Token): Int = a.start
      def end(a: Token): Int = a.end
    }

  implicit val javaToken: TokenOps[JavaToken] =
    new TokenOps[JavaToken] {
      def pos(a: JavaToken): Position = a.pos
      def input(a: JavaToken): Input = a.input
      def start(a: JavaToken): Int = a.start
      def end(a: JavaToken): Int = a.end
    }
}
```

Reply:
Switched to type classes, though now I need to return `TokenDistance[Any]` since we can't have `TokenDistance[Token | JavaToken]` :smiling_face_with_tear: